### PR TITLE
doc: customized auto-instrumentation

### LINKED ISSUE #752: Using custom auto-instrumentation distributions/images
Auto-instrumentation injection is a great feature for people to get started with OpenTelemetry easily ([because of that it's also mentioned in the docs now](https://opentelemetry.io/docs/instrumentation/) 😄  ). 

It would be great if the end-user would have more flexibility on configuring the different injections:

* Add additional instrumentation packages (or extensions) in package.json / requirements.txt or via `-Dotel.javaagent.extensions for java`
* Load a different distribution (my own, vendor-specific) for the `javaagent.jar`, `opentelemetry-distro`, `@opentelemetry/sdk-node`

Would this be an extension to the operator that would be worth pursuing?

### PR DESCRIPTION
**Pull request by cuichenli**
fixes #752

### CODE REVIEW

**Comment by cuichenli** [comment]
i can not add reviewer, so ping @svrnm directly

File: README.md [hunk 1]
```diff
@@ -215,6 +215,60 @@ The possible values for the annotation can be
...
+    argument: "0.25"
+  java:
+    image: your-customized-auto-instrumentation-image:java
+  nodejs:
+    image: your-customized-auto-instrumentation-image:nodejs
+  python:
+    image: your-customized-auto-instrumentation-image:python
+```
+
+To build your customized image, you can start with editing the default dockerfiles. They are stored in the `autoinstrumentation`
```
Comment (svrnm) [inline README.md:246]: > you can start with editing the default dockerfiles.

Replace "with editing" with "by reusing"? Editing could make people assume that they have to commit them back or maintain a forked repository.

File: README.md [hunk 2]
```diff
@@ -215,6 +215,60 @@ The possible values for the annotation can be
...
+
+To build your customized image, you can start with editing the default dockerfiles. They are stored in the `autoinstrumentation` 
+directory. In general, you should comply with the guidelines below.
+- Python 
+  - Ensure the packages are installed in the `/autoinstrumentation` directory. This is required as when instrumenting the pod, 
+    one init container will be created to copy all the content in `/autoinstrumentation` directory to your app's container. Then
+    update the `PYTHONPATH` environment variable accordingly. To achieve this, you can mimic the one in `autoinstrumentation/python/Dockerfile`
+    by using multi-stage builds. In the first stage, install all the required packages in one custom directory with `pip install --target`.
+    Then in the second stage, copy the directory to `/autoinstrumentation`.
+  - Ensure you have `opentelemetry-distro` and `opentelemetry-instrumentation` installed. Those two packages are essential to
```
Comment (svrnm) [inline README.md:254]: Not 100% sure how distributions for python work, could a vendor have a `opentelemetry-<vendor>-distro or `opentelemetry-<vendor>-instrumentation`?

File: README.md [hunk 3]
```diff
@@ -215,6 +215,60 @@ The possible values for the annotation can be
...
+    Python auto-instrumentation.
+  - Grant the necessary access to `/autoinstrumentation` directory. `chmod -R go+r /autoinstrumentation`
+- Node.js
+  - Ensure the packages are installed in the `/autoinstrumentation` directory. This is required as when instrumenting the pod,
+    one init container will be created to copy all the content in `/autoinstrumentation` directory to your app's container. Then
+    update the `NODE_OPTIONS` environment variable accordingly. To achieve this, you can mimic the one in `autoinstrumentation/nodejs/Dockerfile`
+    by using multi-stage builds. In the first stage, install all the required packages in one custom directory.
+    Then in the second stage, copy the directory to `/autoinstrumentation`.
+  - Ensure you have `@opentelemetry/api`, `@opentelemetry/auto-instrumentations-node`, and `@opentelemetry/sdk-node` installed.
```
Comment (svrnm) [inline README.md:263]: Same with Node.JS, not 100% sure here: do we recommend that distributions build on top of those packages or *could* there even be a replacement for them?

**Comment by svrnm** [comment]
Thank you @cuichenli !

File: README.md (hunk 1 above)
Comment (cuichenli) [reply README.md]: updated

File: README.md (hunk 2 above)
Comment (cuichenli) [reply README.md]: My understanding is people can still customize those packages. Updated the wording to suggest it is possible.

File: README.md [hunk 4]
```diff
@@ -215,6 +215,61 @@ The possible values for the annotation can be
...
+
+Although the default auto-instrumentation image can handle some mostly used libraries, it may still 
+miss the ones you need. In this case, you can tell the operator to use your own customized image.
+
+```yaml
+apiVersion: opentelemetry.io/v1alpha1
+kind: Instrumentation
+metadata:
+  name: my-instrumentation
+spec:
```
[the remark this answers is not in the document]
Comment (cuichenli) [reply README.md:228]: updated

File: README.md [hunk 5]
```diff
@@ -215,6 +215,61 @@ The possible values for the annotation can be
...
+    update the `NODE_OPTIONS` environment variable accordingly. To achieve this, you can mimic the one in `autoinstrumentation/nodejs/Dockerfile`
+    by using multi-stage builds. In the first stage, install all the required packages in one custom directory.
+    Then in the second stage, copy the directory to `/autoinstrumentation`.
+  - Ensure you have `@opentelemetry/api`, `@opentelemetry/auto-instrumentations-node`, and `@opentelemetry/sdk-node` or your customized 
+    alternatives installed. 
+  - Grant the necessary access to `/autoinstrumentation` directory. `chmod -R go+r /autoinstrumentation`
+- Java
+  - Download your customized `javaagent.jar` to `/javaagent.jar`.  This is required as when instrumenting the pod,
+    one init container will be created to copy the jar to your app's container.
+  - To customize the instrumentation library, you may need fork https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/javaagent
```
[the remark this answers is not in the document]
Comment (cuichenli) [reply README.md]: removed

File: README.md [hunk 6]
```diff
@@ -215,6 +215,61 @@ The possible values for the annotation can be
 * `"my-instrumentation"` - name of `Instrumentation` CR instance.
 * `"false"` - do not inject
 
+#### Use Customized Image for auto-instrumentation injection
```
[the remark this answers is not in the document]
Comment (cuichenli) [reply README.md]: updated

File: README.md [hunk 7]
```diff
@@ -215,6 +215,61 @@ The possible values for the annotation can be
...
+
+Although the default auto-instrumentation image can handle some mostly used libraries, it may still
```
[the remark this answers is not in the document]
Comment (cuichenli) [reply README.md]: updated

File: README.md [hunk 8]
```diff
@@ -215,6 +215,61 @@ The possible values for the annotation can be
...
+    image: your-customized-auto-instrumentation-image:java
+  nodejs:
+    image: your-customized-auto-instrumentation-image:nodejs
+  python:
+    image: your-customized-auto-instrumentation-image:python
+```
+
+To build your customized image, you can start by reusing the default dockerfiles. They are stored in the `autoinstrumentation` 
+directory. In general, you should comply with the guidelines below.
+- Python
```
[the remark this answers is not in the document]
Comment (cuichenli) [reply README.md]: moved to dockerfiles

File: README.md [hunk 9]
```diff
@@ -215,6 +215,28 @@ The possible values for the annotation can be
...
+spec:
+  java:
+    image: your-customized-auto-instrumentation-image:java
+  nodejs:
+    image: your-customized-auto-instrumentation-image:nodejs
+  python:
+    image: your-customized-auto-instrumentation-image:python
+```
+
+To build your customized image, you can start by reusing the default dockerfiles. They are stored in the `autoinstrumentation`
```
Comment (pavolloffay) [inline README.md:237]: > The Dockerfiles for auto-instrumentation can be found in [autoinstrumentation directory](./autoinstrumentation). Follow the instructions in the Dockerfiles on how to build a custom container image.

Comment (cuichenli) [reply README.md]: updated

**Review by pavolloffay — APPROVED** [review]
Thanks for the PR @cuichenli. This is good enough to go in. We could iterate and improve by making the makefiles more extensible and minimize (e.g. define env vars with URLs for auto-instrumentation libs) the comments in there.

**Pull request merged**